src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Configuration
+{
+    /// <summary>
+    /// Represents a base class for file based <see cref="IConfigurationSource"/>.
+    /// </summary>
+    /// <typeparam name="TProvider">The concrete FileConfigurationProvider to create.</typeparam>
+    public class FileConfigurationSource<TProvider> : FileConfigurationSource where TProvider : FileConfigurationProvider, new()
+    {
+        public override IConfigurationProvider Build(IConfigurationBuilder builder)
```

Comment:
Simpler just to implement Build in the concrete source classes by calling new on the concrete provider classes. You don't need this generic layer at all.